x-pack/legacy/plugins/spaces/server/lib/copy_to_spaces/lib/can_import_into_space.ts

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObjectsClientContract } from 'src/core/server';
+import { SpacesClient } from '../../spaces_client';
+import { CopyToSpaceError } from '../types';
+
+export async function canImportIntoSpace(
+  spaceId: string,
+  spacesClient: SpacesClient,
+  savedObjectsClient: SavedObjectsClientContract
+): Promise<{ canImport: boolean; reason: undefined | CopyToSpaceError['error']['type'] }> {
+  try {
+    const [, canManage] = await Promise.all([
+      spacesClient.get(spaceId),
+      spacesClient.canManageSavedObjects(spaceId),
```

Comment:
> The import API uses the supported types to filter out types that the user wouldn't be authorized to import (if the saved object type is no longer recognized and part of the privileges), and relies on the implicit behavior of us granting access to all saved object types with the saved object management feature.

I recall we made the conscious decision to rely on the saved objects management feature as a simplification measure, to ensure that the user would be sufficiently authorized to perform the action. I'm open to revisiting this decision if needed though.

> The import API currently will work properly, if you only import saved object types which you are authorized to find/bulkGet/bulkCreate. Otherwise, if there is a single type that the user is unauthorized for, it'll fail out part-way through. What do you think about supporting this at the copy-to-space API level as well, and just removing this explicit authorization check?

This would simplify our logic, but I don't love the UX. Designing APIs that partially succeed for opaque reasons is going to lead to a really frustrating experience. We already have to support this to a certain extent with the way that conflicts and missing references are handled today, but I don't want to necessarily let that set a precedent.

> If we don't want to go this route, I'd pedantically like to find a way to prevent us from checking ui:savedObjectsManagement/edit, a ui capability action, to do what we're currently doing, and find a way to "batch" this request instead of doing it for every space.

Sure, let's see what we can come up with. Maybe we could inspect the exported objects, and check to see if the user is authorized to create them in the target spaces?

Reply:
I'm hesitant to rely only on the existing auth checks within the SOC because we still need to be able to generate a list of spaces in the UI for the user to choose from. We can't show them spaces which will fail the subsequent authorization checks

Reply:
What if in the UI itself, we only listed the spaces which the user had the `ui:savedObjectsManagement/edit`, or another dedicated ui action for copying to space. But at the API level, we allowed the user to copy objects without this restriction, as long the types all pass the authorization checks.

A long-long time ago, the import logic in Kibana loaded the entire file into memory client-side and then used the bulkCreate API directly. When we did so, we had the potential to filter out types that the user wasn't authorized to bulkCreate before calling the bulkCreate API itself. When we switched to having a dedicated import/export API, we started submitting the file directly to the API endpoint, and we could no longer rely on the client-side to filter out these types. This exacerbated a few issues:
1) We have no-way server-side in OSS to do any authorization checks explicitly, nor do we expose the saved object types that the user is authorized take action on
2) We have no-way for the security plugin to intercept and alter requests for a specific route. We can use the SOC wrappers to alter the behavior of a specific SOC method, but it felt wrong to add import/export methods to the SOC, as these are one-offs that can reuse the existing SOC methods
3) Security's SOC wrapper will reject any operations that specify any types that the user isn't authorized for. If the user specifies a single type to a find/bulkCreate/bulkGet that the user is unauthorized for, we return a 403. 

I think the proper solution to both import/export and copyToSpace requires us to address one of these true limitations, but I'd like to prevent us doing so at the moment.